Disallow creating instances/disks with snapshots/images from another project

iliana opened the pull request:
Fixes #3335 (@askfongjojo).

This adds these checks:
- in `project_create_disk`: ensure the snapshot belongs to the same project, or ensure the image belongs to the silo or the same project
- in `image_create`: ensure the snapshot belongs to the same project
- in `project_create_instance`: for created disks, implement the same checks as `project_create_disk`

This is a simple check to close off this kind of cross-project resource usage; we may want to enable this kind of thing in the future (perhaps with authz), but I think we will want to be more careful about the design, and it's easier to restrict it now and add functionality to permit this later.

(From when this was a check in the `create_disk` saga: ~~I don't know that where I added these checks is ideal, especially in a [asynchronous control plane operations](https://rfd.shared.oxide.computer/rfd/402) world. It does not seem common to put these checks in at the saga level, possibly for that reason, but I think at minimum having the checks here in addition to possibly adding them before the saga is created is a good idea, to prevent creating new API calls that start a `disk_create` saga without checking these preconditions.~~)

iliana commented:
Realizing now that I missed that you can create an image with a snapshot from another project, adding that too.

zephraph commented:
Woof. Good catch.

zephraph commented:
Hmmm, I see. I'd definitely ask for @davepacheco to weigh in on this as he may have more complex context. My initial reaction is that authz checks should be happening well before a saga initialization. Typically we do all the authz checks at the nexus app layer (and then again at the datastore layer if applicable). Now, if there's code that's doing this that's completely bypassing that code path (e.g. by just directly invoking the saga) then it _would_ bypass the authz checks. Based on the context of the original issue, it looks like this was just some missing checks at the app layer though, so I'm not aware of a need to be extra defensive at the saga level. Again, I could absolutely be missing the bigger picture here so I'd lean on @davepacheco as he as more context on sagas, authz, and their overlap than I do.

iliana commented:
Fortunately the code I added snaps in just fine into the app layer (that's where I originally wrote it before I moved it); means I would need to figure out how to add it to instance create, but at least the test case is written now.

zephraph commented:
I think the place to do the image create check is after this:
https://github.com/oxidecomputer/omicron/blob/c2de4805acfd0c20bef9d87afacac9f8e821ad87/nexus/src/app/image.rs#L216-L220

The `..` portion there will have the silo and project authz objects which we can check against and the parent

zephraph commented:
Hmmm, I'm a bit confused, sorry about this. Looking at @askfongjojo's code example she shows `oxide image create` as doing the wrong thing. Looking at image create I believe that we indeed _don't_ have sufficient checks there. Looks like `disk_create` has that same problem. Yikes.

zephraph commented:
So yeah, we need more checks after this https://github.com/oxidecomputer/omicron/blob/c2de4805acfd0c20bef9d87afacac9f8e821ad87/nexus/src/app/disk.rs#L122-L127

and after this
https://github.com/oxidecomputer/omicron/blob/c2de4805acfd0c20bef9d87afacac9f8e821ad87/nexus/src/app/disk.rs#L146-L150

These are mostly my fault! Sorry about that.

iliana commented:
Yeah that's exactly where I wrote the initial checks. I'll just move 'em and write up the right ones for instance/image creation as well.

iliana commented:
I think in `project_create_instance` we are also not checking everything else that `project_create_disk` checks? e.g. size is a multiple of block size, disk size > snapshot size

iliana commented:
Updated:

- in `project_create_disk`: ensure the snapshot belongs to the same project, or ensure the image belongs to the silo or the same project
- in `image_create`: ensure the snapshot belongs to the same project
- in `project_create_instance`: for created disks, implement the same checks as `project_create_disk`
